kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/ColumnMapping.java

```diff
@@ -260,38 +300,121 @@ private static boolean validModeChange(ColumnMappingMode oldMode, ColumnMappingM
      */
     private static Metadata assignColumnIdAndPhysicalName(Metadata metadata, boolean isNewTable) {
         StructType schema = metadata.getSchema();
-        int maxColumnId = Math.max(
-                Integer.parseInt(metadata.getConfiguration()
-                        .getOrDefault(COLUMN_MAPPING_MAX_COLUMN_ID_KEY, "0")),
-                findMaxColumnId(schema));
+        AtomicInteger maxColumnId = new AtomicInteger(
+                Math.max(
+                        Integer.parseInt(
+                                metadata.getConfiguration().getOrDefault(
+                                        COLUMN_MAPPING_MAX_COLUMN_ID_KEY,
+                                        "0")),
+                        findMaxColumnId(schema)));
         StructType newSchema = new StructType();
         for (StructField field : schema.fields()) {
-            StructField updatedField = field;
-
-            if (!hasColumnId(field)) {
-                maxColumnId++;
-                updatedField = field.withNewMetadata(FieldMetadata.builder()
-                        .fromMetadata(field.getMetadata())
-                        .putLong(COLUMN_MAPPING_ID_KEY, maxColumnId)
-                        .build());
-            }
-            if (!hasPhysicalName(field)) {
-                // re-use old display names as physical names when a table is updated
-                String physicalName = isNewTable ? "col-" + UUID.randomUUID() : field.getName();
-                updatedField = updatedField.withNewMetadata(FieldMetadata.builder()
-                        .fromMetadata(updatedField.getMetadata())
-                        .putString(COLUMN_MAPPING_PHYSICAL_NAME_KEY, physicalName)
-                        .build());
-            }
-            newSchema = newSchema.add(updatedField);
+            newSchema = newSchema.add(
+                    transformAndAssignColumnIdAndPhysicalName(
+                            assignColumnIdAndPhysicalNameToField(field, maxColumnId, isNewTable),
+                            maxColumnId,
+                            isNewTable));
+        }
+
+        if (Boolean.parseBoolean(metadata.getConfiguration()
+                .getOrDefault(TableConfig.ICEBERG_COMPAT_V2_ENABLED.getKey(), "false"))) {
+            newSchema = rewriteFieldIdsForIceberg(newSchema, maxColumnId);
         }
 
         Map<String, String> config = new HashMap<>();
-        config.put(COLUMN_MAPPING_MAX_COLUMN_ID_KEY, Integer.toString(maxColumnId));
+        config.put(COLUMN_MAPPING_MAX_COLUMN_ID_KEY, Integer.toString(maxColumnId.get()));
```

Comment:
`delta-spark` stores the current column id as a [long](https://github.com/delta-io/delta/blob/669dca9c05f1ddb32ce1fd612ffd83eabb1cddd9/spark/src/main/scala/org/apache/spark/sql/delta/DeltaColumnMapping.scala#L409) and fetches it as an [int](https://github.com/delta-io/delta/blob/669dca9c05f1ddb32ce1fd612ffd83eabb1cddd9/spark/src/main/scala/org/apache/spark/sql/delta/DeltaColumnMapping.scala#L158).

The maxColumnId however is stored as a long [here](https://github.com/delta-io/delta/blob/669dca9c05f1ddb32ce1fd612ffd83eabb1cddd9/spark/src/main/scala/org/apache/spark/sql/delta/DeltaColumnMapping.scala#L449) but this is mainly because findMaxColumnId returns a long instead of an int [here](https://github.com/delta-io/delta/blob/669dca9c05f1ddb32ce1fd612ffd83eabb1cddd9/spark/src/main/scala/org/apache/spark/sql/delta/DeltaColumnMapping.scala#L287) (most likely because it fetches the nested column ids as longs). 
I think I mentioned this in an earlier comment on the other PR but I believe the column ID should be an int